Documentation for building websites

mnapoli opened the pull request:
This documentation explains how to build websites using Bref.

What I mean by "website" is a PHP application that also needs to serve static files (via S3).

The goal is to turn this orange rating into green:

<img width="724" alt="Capture d’écran 2019-03-14 à 09 28 17 2" src="https://user-images.githubusercontent.com/720328/54375518-78a02f00-4681-11e9-91dc-93eea1dbceaa.png">

TODO:

- [x] Architecture overview
- [x] S3 documentation
- [x] HTTPS for S3
- [x] CloudFront configuration
- [x] Update the maturity matrix
- [x] Update the text under the maturity matrix

mnapoli commented:
TODO:

- [ ] Disable caching of 404 responses (https://forums.aws.amazon.com/thread.jspa?threadID=220220), this is really painful when developing
- [ ] How to bust CF cache when deploying new changes?

mnapoli commented on docs/websites.md:
````diff
@@ -0,0 +1,191 @@
+                        Resource: !Sub '${Assets.Arn}/*' # everything in the bucket
+```
+
+After [deploying](/docs/deploy.md) the static files will be accessible at `http://<bucket-name>.s3-website-<region>.amazonaws.com`.
````
Oh wow thank you so much for this, I couldn't understand why I was using the `.` notation in my notes… I was so confused by this!

I'll update the docs.

mnapoli commented on docs/websites.md:
````diff
@@ -0,0 +1,191 @@
+aws s3 sync <directory> s3://<bucket-name> --delete --acl public-read
+```
+
+Be aware that `--acl public-read` makes the content of the bucket public!
````
👍 good point, I wasn't setting the policy before so that can be removed indeed.

mnapoli commented:
Thank you @Darkbelg for the review, I've implemented all the changes you suggested!

Darkbelg commented on docs/websites.md:
```diff
@@ -0,0 +1,206 @@
+                    # The assets (S3)
+                    -   Id: Assets
+                        # Watch out, use s3-website URL (https://stackoverflow.com/questions/15309113/amazon-cloudfront-doesnt-respect-my-s3-website-buckets-index-html-rules#15528757)
+                        DomainName: !Sub '${Assets}.s3-website-${AWS::Region}.amazonaws.com'
```
Should also add a note about the structure of the s3 url. https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_website_region_endpoints

mnapoli replied:
Thanks to #324 we can simplify the URL to `https://<bucket>.s3.amazonaws.com/<file>`

kaihendry commented on docs/websites.md:
```diff
@@ -0,0 +1,206 @@
+- API Gateway is *HTTPS only*
+- S3 is *HTTP only*
+
+A simple way to support both HTTP and HTTPS is to use a free CDN like [Cloudflare.com](https://www.cloudflare.com/).
```
Wouldn't AWS CloudFront make a better suggestion for providing HTTPS support on a bucket?

mnapoli replied:
CloudFront would definitely work and be a good option, but it is a bit complex. Cloudflare has a very simple UI: you basically sign up, add the website and setup your DNS and that's it (and it's free).

I'm not sure about suggesting Cloudflare though, it might be more confusing than anything for those who don't know about it 🤔

mnapoli replied:
I've removed mentions of Cloudflare, indeed it is better to stay with AWS and use CloudFront.

kaihendry commented:
Sorry, little confused. Is SAM's template.yaml sufficient for orchestrating the entire CloudFront distribution?

I'm always confused when I should be using **terraform** or **cloudformation** for these sorts of tasks.

Be good to see a complete example since I find piecing together YAML files really frustrating, esp when it comes to its crazy indentation.

Thank you, great work btw :ok_hand:

mnapoli commented:
@kaihendry yes it is. From my understanding Terraform is like CloudFormation except it supports more than just AWS. If you use just AWS then CloudFormation is enough.

If you are using PhpStorm be sure to install the CloudFormation plugin, it makes dealing with the YAML much easier (it validates what you type). As for a complete example that's a good idea, maybe a separate repository!

By the way a quick update here (this PR is not dead at all), I may have found a solution to turn all the YAML shown here into this:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:
    # define your functions here as usual
    ...

    WebsiteCDN:
        Type: AWS::CloudFormation::Stack
        Properties:
            TemplateURL: vendor/mnapoli/bref/cloudformation/website.yaml
            Parameters:
                ServerlessRestApi: !Ref ServerlessRestApi
                AssetsBucketName: <s3-bucket-name>
                DomainName: example.com
                CertificateArn: arn:aws:acm:us-east-1:216536346254:certificate/322f12ee-1165-4bfa-a41f-08c932a2935d
```

That will be much simpler, though limited of course to simple use cases. Still working on it.

chrisemerson commented:
For me, I use Terraform to manage the rest of my site, because frankly it's a lot nicer than Cloudformation. I'm just playing now for the first time but having support for Terraform as well as CF might be nice, though they may play nicely enough together as it stands. Supporting more than AWS isn't the only difference though.

I noticed also you have a note on "How to bust CF cache when deploying new changes?" - if CF here refers to CloudFront rather than Cloudflare, the way to do it is to create an "invalidation" - this can be done via the AWS CLI and if you create one for * it'll do a whole CF distribution, usually takes a minute or 2 to take effect.

mnapoli commented on docs/websites.md:
```diff
@@ -0,0 +1,206 @@
+            DistributionConfig:
+                Enabled: true
+                # Cheapest option by default (https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_DistributionConfig.html)
+                PriceClass: PriceClass_100
```
Is it still compatible with HTTP 1?

cmaerz replied:
![image](https://user-images.githubusercontent.com/1479186/59347603-6338b080-8d15-11e9-95f4-71acb6db27f1.png)
That's how it looks in AWS Backend after deployment so it looks like it.

mnapoli replied:
Perfect thanks!

mnapoli commented on docs/websites.md:
```diff
@@ -0,0 +1,206 @@
+                    # ARN of the certificate created in ACM
+                    AcmCertificateArn: <certificate-arn>
+                    # See https://docs.aws.amazon.com/fr_fr/cloudfront/latest/APIReference/API_ViewerCertificate.html
+                    SslSupportMethod: 'sni-only'
```
Does that mean that some browsers/clients will not be supported?

cmaerz replied:
TLSv1 is considered "unsafe" so it doesn't even make sense to support them. (See screenshot above)

deleugpn replied:
Earlier versions of TLS has reached end of life. Any browser not supporting at least TLS 1.1 is probably EOL as well.

mnapoli replied:
Perfect, since it's recommended by AWS as well it's a safe bet.

Thank you!

cmaerz commented:
Maybe we should also document how to get the real Hostname through Cloudfront with a Lambda@edge function as you need it in some applications:

- Create a new Lambda Function with Node 8.10 at `US East (N. Virginia)`
- Paste the following code:

```
'use strict';
 
const static_host_header = '<ARN OF YOUR WEBSITE LAMBDA FUNCTION HERE>'; 
 
exports.handler = (event, context, callback) => {
    const request = event.Records[0].cf.request;
    const headers = request.headers;
    
    headers['x-forwarded-host'] = [ { 
        key: 'X-Forwarded-Host', 
        value: headers.host[0].value 
    }];
 
    headers['host'] = [ { 
        key: 'Host',
        value: static_host_header
    } ];
 
    callback(null, request);
};
```

**Note:** I copied the code above somewhere from some forum and modified it a bit. Credits go there but I can't find the source currently
- Setup a Lambda execution Role for the function (needs to be documented.. tbh i did this a few weeks ago and i have no idea how I did it :D)
```
WebsiteCDN:
    Type: AWS::CloudFront::Distribution
    Properties:
      DistributionConfig:
      ...
       LambdaFunctionAssociations:
            - EventType: 'origin-request'
              LambdaFunctionARN: '<YOUR ARN HERE>'
```
- Use `$_SERVER['HTTP_X_FORWARDED_HOST']` to get the hostname the user requested in the first place

mnapoli commented:
@cmaerz I did a similar behavior but with a `viewer-request` event, which avoids to hardcode the API Gateway URL in the code. Here is the code I used:

```js
exports.handler = (event, context, callback) => {
  const request = event.Records[0].cf.request
  request.headers['x-forwarded-host'] = [
    { key: 'X-Forwarded-Host', value: request.headers.host[0].value }
  ]
  return callback(null, request)
}
```

The downsides I see:

- this lambda will be executed for every request, including assets or cached requests (i.e. costs and latency)
- viewer-request events require that lambdas use `128M` max of memory, meaning the execution is (much) slower than a 1024M origin-request event (maybe I'm wrong about this)

The advantages:

- no need to change any configuration, it's compatible with any CloudFront setup
- simpler code since it's generic

mnapoli commented:
Will fix #324

mnapoli commented:
This is not a WIP anymore, reviews are welcome.

Changes since last time:

- I have rewritten the documentation to match with the `serverless.yml` syntax
- no mention of CloudFlare anymore (in favor of CloudFront)
- simplified S3 URLs
